File: project.py
from json import JSONEncoder, loads
from common import JSONable, check_duplicate_id
from ProjectError import InvalidIdException, InvalidTypeException, \
                         MissingFieldException, InvalidProgressException, \
                         DuplicateIdException


class Status(JSONable):
        STATUS_CODES_MAJOR = {
                0: 'Not started',
                1: 'In progress',
                2: 'Completed',
                -1: 'Unknown status',
        }
        STATUS_CODES_MINOR = {
                0: None,
                1: 'Awaiting dependencies',
        }

        def __init__(self, major, minor = 0):
                self.major = major
                self.minor = minor

        @property
        def major_code(self):
                return self.STATUS_CODES_MAJOR[self.major]

        @property
        def minor_code(self):
                if self.minor != 0:
                        return self.STATUS_CODES_MINOR[self.minor]
                else:
                        return self.major_code

        def __str__(self):
                return '({}, {}): {}'.format(self.major,
                                             self.minor,
                                             self.minor_code)

        def __eq__(self, o):
                if self.major == o.major and self.minor == o.minor:
                        return True
                else:
                        return False


MAX_STATUS = 2
PROGRESS_STATUS = 1
CURRENT_ID = None

PROPS = ['name', 'desc', 'required', 'progress', 'meta', 'counter',
         'dependencies']
TYPES = {
        'id': int,
        'name': str,
        'desc': str,
        'required': int,
        'progress': int,
        'meta': int,
        'counter': bool,
        'dependencies': list
}


def get_id():
        global CURRENT_ID
        temp = CURRENT_ID
        CURRENT_ID += 1
        return temp


class Project(JSONable):
        def __init__(self, system, name, required = MAX_STATUS, id = None,
                     meta = 0, counter = False):
                self.name = name
                self.required = required
                self.progress = 0
                self.system = system
                if id is None:
                        self.id = get_id()
                else:
                        self.id = id
                self.dependencies = []
                self.meta = meta
                self.desc = ''
                self.successors = []
                self.counter = counter
                self._status = Status(0, 0)

        @property
        def status(self):
                return self._status

        def update_status(self):
                '''Recalculate the current status, returns whether the status changed'''
                old_status = self._status
                self._status = self._recalculate_status()
                if self._status != old_status:
                        return True
                else:
                        return False

        def _recalculate_status(self):
                if self.meta == 0:
                        if self.progress == 0:
                                depend_status = self._check_depends()
                                if depend_status.major != MAX_STATUS:
                                        return Status(0, 1)
                                else:
                                        return Status(0, 0)
                        elif self.progress < self.required:
                                return Status(PROGRESS_STATUS, 0)
                        else:
                                return Status(MAX_STATUS, 0)
                elif self.meta == 1:
                        return self._check_depends()
                else:
                        return -1

        def _check_depends(self):
                progress = 0
                completed = True
                for depend in self.dependencies:
                        project = self.system.get_event_by_id(depend)
                        if project.status.major != MAX_STATUS:
                                completed = False
                        progress = max(progress, project.status.major)
                if completed:
                        return Status(MAX_STATUS)
                elif progress > 0:
                        return Status(PROGRESS_STATUS)
                else:
                        return Status(0)

        excludes = set(['system'])
        includes = set(['status'])

        def json(self):
                result = super().json(self.excludes, self.includes)
                result['type'] = 'Project'
                result['status'] = self.status
                return result

        @classmethod
        def fromJSON(cls, json, system):
                obj = loads(json)
                return cls.fromJSONObj(obj, system)

        @classmethod
        def fromJSONObj(cls, obj, system):
                cls._validate_obj(obj, system)
                name, required, = obj['name'], obj['required']
                meta = obj['meta']
                counter = obj['counter']
                result = cls(system, name, required, meta = meta,
                             counter = counter)
                result.progress = obj['progress']
                result.dependencies = obj['dependencies']
                result.desc = obj['desc']
                result.update_status()
                return result

        @classmethod
        def _validate_obj(cls, obj, system):
                for prop in PROPS:
                        if prop not in obj:
                                raise MissingFieldException(prop, 'Project')
                for prop in PROPS:
                        if type(obj[prop]) is not TYPES[prop]:
                                raise InvalidTypeException(
                                        type(obj[prop]),
                                        prop,
                                        'Project',
                                        TYPES[prop])
                for id in obj['dependencies']:
                        if type(id) is not int:
                                raise InvalidTypeException(
                                        type(id),
                                        'id',
                                        'Project.dependencies',
                                        int)
                if obj['required'] < 1:
                        raise InvalidProgressException(obj['required'],
                                                       'required')
                if obj['required'] < obj['progress']:
                        raise InvalidProgressException(obj['progress'],
                                                       'progress')
                if not obj['counter'] and obj['required'] != 2:
                        raise InvalidProgressException(obj['required'],
                                                       'required')

                if obj['progress'] < 0:
                        raise InvalidProgressException(obj['required'],
                                                       'progress')

                for id in obj['dependencies']:
                        if id not in system.projects:
                                raise InvalidIdException(id)

                duplicate = check_duplicate_id(obj['dependencies'])
                if duplicate:
                        raise DuplicateIdException(id, 'dependencies')

        def add_dependency(self, id: int):
                if id in self.dependencies:
                        raise DuplicateIdException(id, 'project.dependencies')
                if id not in self.system.projects:
                        raise InvalidIdException(id)
                self.dependencies.append(id)
                parent = self.system.get_event_by_id(id)
                parent.successors.append(id)

        def remove_dependency(self, id: int):
                if id not in self.dependencies:
                        raise InvalidIdException(id)
                self.dependencies.remove(id)
                parent = self.system.get_event_by_id(id)
                parent.successors.remove(id)


class ProjectEncoder(JSONEncoder):
        def default(self, o):
                try:
                        return o.json()
                except AttributeError:
                        super().default(o)

File: projects/Project.js
'use strict'

/**
 * @event Projects.Project#update
 * @type {Object}
 * @property {String} target Name of property changed
 */

/**
 * Interface for display of a Projects.Project
 * @interface ProjectDisplay
 */
/**
 * @function update
 * @description Fired upon a change event
 * @name ProjectDisplay.update
 * @param {Projects.Project#update} event Event been fired
 */

/**
 * Project namespace
 * @namespace Projects
 */
const Projects = {
	ProjectParseError: class extends Error {
		constructor (...args) {
			super(...args);
			Error.captureStackTrace(this, this.constructor);
		}
	},
	STATUS_CODES_MAJOR: {
		0: 'Not started',
		1: 'In progress',
		2: 'Completed',
		'-1': 'Unknown status',
	},
	STATUS_CODES_MINOR: {
		0: null,
		1: 'Awaiting dependencies',
	},
	MAX_STATUS: 2,
	PROGRESS_STATUS: 1,
	Status: class Status {
		constructor (major, minor = 0) {
			this.major = major;
			this.minor = minor;
		}
		get major_code () {
			return Projects.STATUS_CODES_MAJOR[this.major];
		}
		get minor_code () {
			if (this.minor !== 0) {
				return Projects.STATUS_CODES_MINOR[this.minor];
			} else {
				return this.major_code;
			}
		}
		static fromJSONObj (obj) {
			let status = new this(obj.major, obj.minor);
			return status;
		}
	},
	Project: class Project {
		constructor (system, name, id, desc = '', required = 2, status = null) {
			this._system = system;
			this._name = name;
			this._desc = desc;
			this._dependencies = [];
			this.id = id;
			this._required = required;
			this._progress = 0;
			this._meta = 0;
			if (status !== null) {
				this._status = status;
			} else {
				this._status = new Projects.Status(0);
			}

			this._displays = new Set();
			this.type = 'Project';
		}
		get name () {
			return this._name;
		}
		set name (value) {
			this._name = value;
			this.dispatchUpdate();
		}
		get desc () {
			return this._desc;
		}
		set desc (value) {
			this._desc = value;
			this.dispatchUpdate();
		}
		get dependencies () {
			return this._dependencies;
		}
		set dependencies (value) {
			this._dependencies = value;
			this.dispatchUpdate();
		}
		get required () {
			return this._required;
		}
		set required (value) {
			this._required = value;
			this.dispatchUpdate();
		}
		get progress () {
			return this._progress;
		}
		set progress (value) {
			this._progress = value;
			this.dispatchUpdate();
		}
		get meta () {
			return this._meta;
		}
		set meta (value) {
			this._meta = value;
			this.dispatchUpdate();
		}
		dispatchUpdate () {

		}
		addDisplay (display) {
			this._displays.add(display);
		}
		removeDisplay (display) {
			this._displays.delete(display);
		}
		get status () {
			return this._status;
		}
		set status (value) {
			this._status = value;
			this.dispatchUpdate();
		}
		get status_code () {
			return this.status.minor_code;
		}
		static fromJSONObj(obj, system) {
			if (obj.type !== 'Project') {
				throw new Projects.ProjectParseError('Not a Project representation');
			}
			let status = Projects.Status.fromJSONObj(obj.status);
			let project = new this(system, obj.name, obj.id, obj.desc, obj.required, status);
			project.dependencies = obj.dependencies;
			project.progress = obj.progress;
			project.meta = obj.meta;
			return project
		}
		toJSON () {
			return Elements.jsonIncludes(this, this.constructor.json_props);
		}
		static get json_props () {
			return ['name', 'desc', 'dependencies', 'required', 'progress', 'meta'];
		}
	},
	System: class System {
		constructor() {
			this.projects = new Map();
		}
		static fromJSON(json) {
			return this.fromJSONObj(JSON.parse(json))
		}
		static fromJSONObj(obj) {
			if (obj.type !== 'System') {
				throw new Projects.ProjectParseError('Not a System');
			}
			let system = new Projects.System();
			for (let project_key in obj.projects) {
				let projectObj = obj.projects[project_key]
				let project = Projects.Project.fromJSONObj(projectObj, system);
				system.projects.set(project.id, project);
			}
			return system;
		}
		get_event_by_id (id) {
			return this.projects.get(id);
		}
		async add_project (project) {
			let message = JSON.stringify([project]);
			console.log('Sending: ', message);
			let form_data = new FormData();
			form_data.append('create', message);
			let fetch_promise;
			try {
				fetch_promise =  fetch(window.location + 'create', {
					method: 'POST',
					body: form_data,
				});
			} catch (e) {
				alert('Failed to connect to server');
				throw e;
			}
			try {
				this.patch(fetch_promise);
				return true;
			} catch (e) {
				return false;
			}
		}
		async patch (promise) {
			let patch;
			try {
				let response = await promise;
				patch = await response.json();
			} catch (e) {
				alert('Bad response from server');
				throw e;
			}
			console.log ('Would apply patch: ', patch);
			for (let create of patch.create) {
				this._patch_add_project(create);
			}
		}
		async _patch_add_project (project_obj) {
			let project = Projects.Project.fromJSONObj(project_obj, this);
			this.projects.set(project.id, project);
			// TODO: Move this into own element
			let projects = document.querySelector('#projects');
			let display = document.createElement('elements-projects-project-display');
			display.data = project;
			projects.append(display);

		}

	}
}

Elements.loaded('projects-Project');

const test = () => {
	const base1 = {
		"type":"Project",
		"name":"a",
		"desc":"aa",
		"dependencies":[],
		"required":2,
		"progress":0,
		"meta":1,
		"status": {
			"major": 2,
			"minor": 0,
		},
	};
	const base2 = {
		"type":"Project",
		"name":"b",
		"desc":"bb",
		"dependencies":[],
		"required":2,
		"progress":0,
		"meta":0,
		"status": {
			"major": 0,
			"minor": 0,
		},
	};
	const compare_project = (proj1, proj2) => {
		const props = ['name', 'desc', 'required', 'progress', 'meta'];
		for (let prop of props) {
			if (proj1[prop] !== proj2[prop]) {
				console.log('Mismatched property ', prop);
				return false;
			}
		}
		// Check depends
		let [set1, set2] = [new Set(proj1.dependencies), new Set(proj2.dependencies)];
		if (set1.size !== proj1.dependencies.length) {
			console.log('Duplicate dependencies');
			return false;
		}
		if (set2.size !== proj2.dependencies.length) {
			console.log('Duplicate dependencies');
			return false;
		}
		if (set1.size !== set2.size) {
			console.log('Mismatched dependencies');
		}
		for (let depend of set1) {
			if (!set2.has(depend)) {
				console.log('Mismatched dependencies');
				return false;
			}
		}
		return true;
	};
	let promises = [];
	for (let test of [base1, base2]) {
		let resolve_f, reject_f;
		let promise = new Promise((resolve, reject) => {
			[resolve_f, reject_f] = [resolve, reject];
		});
		promises.push(promise);
		const data1 = new Projects.System();
		let project_js = Projects.Project.fromJSONObj(test, data1);

		data1._patch_add_project = async (project_obj) => {
			let project_py = Projects.Project.fromJSONObj(project_obj, data1);
			if (!compare_project(project_js, project_py)) {
				reject_f();
			} else {
				resolve_f();
			}
		};
		data1.add_project(project_js);
	}
	(async () => {
		try {
			await Promise.all(promises);
			console.log('Passed tests');
		} catch (e) {
			console.log('Failed tests');
			throw e;
		}
	})();
}
